File: parse_dsalist.py
#!/usr/bin/env python

import sys
import re
from json import JSONEncoder

REGEX_1ST_LINE = r"^\[(?P<date>[^]]+)\]\s*(?P<ident>[A-z0-9\-]+)\s*(?P<package_name>\S+)\s*-*\s*(?P<typ>.*)$"
REGEX_CVE_LINE = r"\s+{(?P<cves>[^}]*)}"
REGEX_REL_LINE = r"\s+\[(?P<release>[^\]]*)\]\s*-\s*(?P<package_name>\S+)\s*(?P<version>\S*)"
REGEX_NOT_LINE = r"\s+NOTE:"

re_1st = re.compile(REGEX_1ST_LINE)
re_cve = re.compile(REGEX_CVE_LINE)
re_rel = re.compile(REGEX_REL_LINE)
re_not = re.compile(REGEX_NOT_LINE)


class JSONDSAEncoder(JSONEncoder):
    def default(self, o):
        if(isinstance(o, DSA)):
            return o.dict()
        else:
            JSONEncoder.default(o)


class ParserException(RuntimeError):
    def __init__(self, lnum=-1, line=None, msg="ParserException", critical=True):
        RuntimeError.__init__(self, "at %d: %s (%r)" % (lnum, msg, line))
        self.lnum = lnum
        self.line = line
        self.critical = critical


class ParserError(ParserException):
    def __init__(self, *args, **kwargs):
        kwargs['critical'] = True
        ParserException.__init__(self, *args, **kwargs)


class ParserWarning(ParserException):
    def __init__(self, *args, **kwargs):
        kwargs['critical'] = False
        ParserException.__init__(self, *args, **kwargs)


class DSA():
    def __init__(self, date=None, ident=None, typ=None, package_name=None):
        self.date = date
        self.id = ident
        self.type = typ
        self.package_name = package_name
        self.versions = {}
        self.cve = []

    def add_cve(self, cve_numbers):
        self.cve = list(cve_numbers)

    def add_release(self, release, package_name, version):
        if(release not in self.versions):
            self.versions[release] = {}
        self.versions[release][package_name] = version

    def pp(self):
        print("%s from %s for %s" % (self.id, self.date, self.package_name))
        print("  CVE: %r" % self.cve)
        for rel in self.versions.keys():
            print("\t%s" % rel)
            rel_dict = self.versions[rel]
            for p in rel_dict.keys():
                print("\t\t%s %s" % (p, rel_dict[p]))

    def dict(self):
        return {
                'name': self.id,
                'date': self.date,
                'type': self.type,
                'package': self.package_name,
                'cve': self.cve,
                'versions': self.versions
                }


def parse_dsa_list(path):
    dsa = None
    dsa_list = []
    i = 0

    with open(path, "rw") as f:
        for line in f:
            try:
                i += 1
                res1 = re_1st.match(line)
                if(res1):
                    if(dsa):
                        dsa_list.append(dsa)
                    dsa = DSA(**res1.groupdict())
                elif(dsa):
                    res = re_rel.match(line)
                    if(res):
                        dsa.add_release(**res.groupdict())
                        continue
                    res = re_cve.match(line)
                    if(res):
                        if(res.lastindex > 0):
                            dsa.add_cve(res.group('cves').split(' '))
                        continue
                    if(re_not.match(line)):
                        # ignore 'NOTE:' lines
                        continue

                    raise ParserWarning(lnum=i, line=line, msg='Unknown Line in DSA')

                else:
                    raise ParserWarning(lnum=i, line=line, msg='Unknown Line')
            except ParserException as e:
                if(e.critical):
                    raise
                else:
                    sys.stderr.write("%s\n" % e)

    if(dsa):
        dsa_list.append(dsa)

    return dsa_list


if __name__ == "__main__":
    import json

    dsa_list = parse_dsa_list("test_data/dsa.list")

    print(json.dumps(dsa_list, cls=JSONDSAEncoder))

    #for dsa in dsa_list:
    #    dsa.pp()

File: parse_dsalist.rb
# frozen_string_literal: true

require 'json'

REGEX_1ST_LINE = /^\[(?<date>[^\]]+)\]\s*(?<ident>[A-z0-9-]+)\s*(?<package>\S+)\s*-*\s*(?<typ>.*)$/
REGEX_CVE_LINE = /\s+{(?<cves>[^}]*)}/
REGEX_REL_LINE = /\s+\[(?<release>[^\]]*)\]\s*-\s*(?<package>\S+)\s*(?<version>\S*)/
REGEX_NOT_LINE = /\s+NOTE:/

# Base-Parser Exception
class ParserException < RuntimeError
  def initialize(lnum = -1, line = nil, msg = 'ParserException', critical: true)
    super("at #{lnum}: #{msg} (#{line})")
    @lnum = lnum
    @line = line
    @critical = critical
  end
end

# critical Parser-Error
class ParserError < ParserException
  def initialize(lnum, line, msg)
    super
    @critical = true
  end
end

# uncritical ParserError
class ParserWarning < ParserException
  def initialize(lnum, line, msg)
    super
    @critical = false
  end
end

# {'a'=>1, 'b'=>2} to {:a=>1, :b=>2}
# keys must be symbols for usage as named method-parameters
def symbolize_hash_keys(hsh)
  return nil if hsh.nil?

  res = {}
  hsh.each do |k, v|
    res[k.to_sym] = v
  end
  res
end

# "smart" list of DSAs
class DSAList < Array
  def initialize
    @opt_ignore_empty_cve = true
    super
  end

  def secure_push(item)
    return unless item.is_a? DSA
    return if @opt_ignore_empty_cve && item.cve_empty?

    push item
  end
end

# DebianSecurityAdvisory
class DSA
  PARSE_DSA_LIST_OPTIONS_DEFAULTS = {
    ignore_empty_cve: true
  }.freeze

  attr_reader :date, :id, :type, :package, :versions, :cve

  def initialize(date:, ident:, typ:, package:)
    @date = date
    @id = ident
    @type = typ
    @package = package
    @versions = {}
    @cve = []
  end

  def add_cve(cve_numbers)
    @cve = Array.new(cve_numbers)
  end

  def cve_empty?
    (@cve.nil? || @cve.empty?)
  end

  def add_release(release:, package:, version:)
    @versions[release] = {} unless @versions.key? release
    @versions[release][package] = version
  end

  def pp
    puts("#{@id} from #{@date} for #{@package}")
    puts("  CVE: #{@cve}")
    @versions.each_key do |rel|
      puts("\t#{rel}")
      rel_dict = @versions[rel]
      rel_dict.each_key do |p|
        puts("\t\t#{p} #{rel_dict[p]}")
      end
    end
  end

  # for conversion to JSON
  def to_h
    {
      'name' => @id,
      'date' => @date,
      'type' => @type,
      'package' => @package,
      'cve' => @cve,
      'versions' => @versions
    }
  end

  def to_json(*options)
    to_h.to_json(*options)
  end

  # Recreate data from original DSA-list file
  # param release_filter Array of relase_name String, e.g. `[ 'bullseye', 'stretch' ]`
  # returns String e.g.:
  # [11 Dec 2021] DSA-5020-1 apache-log4j2 - security update
  # \t{CVE-2021-44228}
  # \t[buster] - apache-log4j2 2.15.0-1~deb10u1
  # \t[bullseye] - apache-log4j2 2.15.0-1~deb11u1
  def to_orig(release_filter=nil)
    e = StringIO.new
    e.puts "[#{@date}] #{@id} #{@package} - #{@type}"
    e.puts "\t{#{@cve.join(' ')}}"

    versions = @versions
    # filter releases and return nil if no remaining releases
    if release_filter.respond_to? :include?
      versions = versions.select { |v| release_filter.include?(v) }
      return nil if versions.empty?
    end

    versions.each do |rel, packages|
      packages.each do |name, version|
        e.puts "\t[#{rel}] - #{name} #{version}"
      end
    end
    e.string
  end

  ## Class methods expects IO
  def self.parse_dsa_list_str(string)
    parse_dsa_list(StringIO.new(string))
  end

  def self.parse_dsa_list(io)
    raise "requires IO object; got #{io.class.name}" unless io.respond_to? :each_line

    dsa = nil
    dsa_list = DSAList.new
    i = 0

    io.each_line do |line|
      i += 1
      res1 = REGEX_1ST_LINE.match(line)
      if res1
        dsa_list.secure_push dsa
        dsa = DSA.new(**symbolize_hash_keys(res1.named_captures))
      elsif dsa
        res = REGEX_REL_LINE.match(line)
        if res
          dsa.add_release(**symbolize_hash_keys(res.named_captures))
          next
        end
        res = REGEX_CVE_LINE.match(line)
        if res
          dsa.add_cve(res[:cves].split) unless res[:cves].empty?
          next
        end
        if REGEX_NOT_LINE.match(line)
          # ignore 'NOTE:' lines
          next
        end

        raise ParserWarning.new(i, line, 'Unknown Line in DSA')

      else
        raise ParserWarning.new(i, line, 'Unknown Line')
      end
    rescue ParserException => e
      raise if e.critical

      warn(e)
    end

    dsa_list.secure_push dsa
  end
end

if $PROGRAM_NAME == __FILE__
  dsa_list = nil
  File.open('test/data/dsa.list', 'r') do |f|
    dsa_list = DSA.parse_dsa_list(f)
  end

  if ENV['FILTERED_ORIGINAL']
    release_whitelist = ['bullseye']
    dsa_list.each do |dsa|
      s = dsa.to_orig(release_whitelist)
      puts s unless s.nil?
    end
  else
    puts JSON.generate(dsa_list)
  end
end
